feat(List): add aesop / simp annotations to selected lemmas for improved automation

farruhx opened the pull request:
This PR adds `@[aesop safe]` and `@[simp]` annotations to a set
of List lemmas whose proofs are routine and benefit from standardized
automation. No statements or definitions are changed; only proof annotations
are added.

The lemmas updated in this PR are:

* `or_exists_of_exists_mem_cons`
* `append_subset_of_subset_of_subset`
* `map_subset_iff`
* `append_eq_has_append`
* `append_right_injective`
* `append_left_injective`
* `reverse_surjective`
* `reverse_bijective`
* `mem_getLast?_append_of_mem_getLast?`
* `mem_dropLast_of_mem_of_ne_getLast`
* `idxOf_eq_length_iff`
* `idxOf_append_of_mem`
* `length_eraseP_add_one`

The goal is to make these commonly used lemmas easier for `aesop`-based
automation to resolve, while avoiding any interference with simp-normal-form
lemmas or canonical rewrite rules.

There are no API changes and no new theorems—only improved automation behavior.

farruhx committed:
feat: improve automation for List and Finset lemmas

This commit annotates and refactors several lemmas (A1, A3, A4, A5, A6, A7, A8)
to use , , or strengthened  lemmas when appropriate.

euprunin commented:
Is this the correct version? This PR doesn't appear to contain any `grind` annotations, even though the title suggests that it should.

farruhx committed:
Removed @[simp] annotation due to lint error. Will follow up with further fix for this in the subsequent PRs

farruhx committed:
Removed @[simp] from exists_mem_cons_iff due to the lint error

farruhx committed:
Merge branch 'master' into feature/simplify-basics-with-aesop

Commit:
feat: add basic List lemmas with Aesop annotations and simp normal forms

This commit adds a collection of small but useful lemmas for ,
filling several gaps in the API around , , ,
, and .  These lemmas are elementary but appear
frequently in proofs and help both  and  reason more
smoothly about lists.

New lemmas include:

*  and  with safe Aesop
  rules and corresponding  ↔ lemmas.
*  and , exposed as safe facts
  for Aesop.
* Membership and structural lemmas for , , and
  concatenation:
  -
  -
* Length and index lemmas:
  -  as a  normalization lemma
  - directional Aesop lemma
  -
  -

All Aesop annotations have been chosen conservatively as
to avoid over-aggressive normalization while still enabling automated
forward reasoning.  Lemmas with  conclusions are provided as
in accordance with mathlib style and to satisfy the  linter.

These additions improve automation coverage, reduce boilerplate in
typical list proofs, and help Aesop discharge routine goals that
previously required manual steps.

farruhx committed:
Added missing members

farruhx committed:
Merge branch 'master' into feature/simplify-basics-with-aesop

farruhx committed:
Remove append_left_injective' and append_right_injective' due to lint issue

artie2000 commented on Mathlib/Data/List/Basic.lean:
```diff
@@ -171,10 +172,12 @@ theorem cons_subset_of_subset_of_mem {a : α} {l m : List α}
   cons_subset.2 ⟨ainm, lsubm⟩
 
+@[aesop safe]
 theorem append_subset_of_subset_of_subset {l₁ l₂ l : List α} (l₁subl : l₁ ⊆ l) (l₂subl : l₂ ⊆ l) :
```
This doesn't seem like a good `safe` rule. Can you justify that applying it is provability-preserving?

artie2000 commented on Mathlib/Data/List/Basic.lean:
```diff
@@ -367,9 +375,11 @@ theorem getLast?_append_of_ne_nil (l₁ : List α) :
 
+@[aesop safe]
 theorem mem_dropLast_of_mem_of_ne_getLast {a : α} (ha : a ∈ l)
     (ha' : a ≠ l.getLast (ne_nil_of_mem ha)) : a ∈ l.dropLast := by
```
This is not provability-preserving. What if `l=[x,x]`? This should be a high-priority `unsafe` rule.

artie2000 commented on Mathlib/Data/List/Basic.lean:
```diff
@@ -526,9 +536,14 @@ theorem idxOf_cons_eq {a b : α} (l : List α) : b = a → idxOf a (b :: l) = 0
+@[aesop safe]
+theorem not_mem_of_idxOf_eq_length {a : α} {l : List α} (h : idxOf a l = length l) : a ∉ l :=
+  (idxOf_eq_length_iff.mp h)
+
```
This sets up an infinite loop with your previous rule. You need to pick one of `idxOf a l = length l` and `a ∉ l` as your `aesop` normal form.

artie2000 reviewed (commented):
I am a bit dubious about some of these `aesop safe` rules. `aesop safe` rules should preserve provability under the ruleset.
